File: src/localization/localization/global_localizer_node.py
#!/usr/bin/env python3

"""
ROS2 global localizer node (Complete Version)

This node implements Monte Carlo Localization (MCL) using a Particle Filter.
It subscribes to map, scan, and tf to localize the robot in the map frame.

Flow:
1. Initialize Particle Filter with given parameters.
2. Receive Map -> PF.set_map()
3. Receive Scan:
    a. Calculate motion delta from Odom TF (Prediction) -> PF.predict()
    b. Update particles with LaserScan (Correction) -> PF.update()
    c. Resample particles -> PF.resample()
    d. Calculate Map->Odom transform
    e. Publish Map->Odom TF and visualization topics
"""

import rclpy
from rclpy.node import Node
from rclpy.time import Time
import math

from geometry_msgs.msg import TransformStamped, PoseWithCovarianceStamped, PoseArray, Pose, PoseStamped
from nav_msgs.msg import OccupancyGrid
from sensor_msgs.msg import LaserScan
from tf2_ros import TransformBroadcaster, Buffer, TransformListener, LookupException, ConnectivityException, \
    ExtrapolationException
import tf_transformations
from rclpy.qos import QoSProfile, ReliabilityPolicy, DurabilityPolicy

import numpy as np

# 사용자 정의 모듈 임포트
from utils import transform_to_matrix
from particle_filter import ParticleFilter

# 3D 오도메트리 매트릭스에서 2D(x, y, yaw) 매트릭스만 추출하는 함수
def get_2d_matrix(matrix_4x4):
    x = matrix_4x4[0, 3]
    y = matrix_4x4[1, 3]
    # 회전에서 Yaw만 추출
    _, _, yaw = tf_transformations.euler_from_matrix(matrix_4x4)

    # 새로운 2D 기반 4x4 행렬 생성 (Z, Roll, Pitch 제거)
    q = tf_transformations.quaternion_from_euler(0, 0, yaw)
    new_mat = tf_transformations.quaternion_matrix(q)
    new_mat[0, 3] = x
    new_mat[1, 3] = y
    return new_mat


class GlobalLocalizerNode(Node):
    def __init__(self):
        super().__init__('global_localizer')
        self.get_logger().info('Global localizer (MCL) node initialized')

        # Parameters
        self.declare_parameter('initial_pose_x', 0.0)
        self.declare_parameter('initial_pose_y', 1.0)  # Default y=1
        self.declare_parameter('initial_pose_z', 0.5)  # Default z=0.5
        self.declare_parameter('initial_pose_roll', 0.0)
        self.declare_parameter('initial_pose_pitch', 0.0)
        self.declare_parameter('initial_pose_yaw', 0.0)
        self.declare_parameter('min_particles', 300)
        self.declare_parameter('max_particles', 1000)

        self.init_x = self.get_parameter('initial_pose_x').value
        self.init_y = self.get_parameter('initial_pose_y').value
        self.init_z = self.get_parameter('initial_pose_z').value  # 저장 필요
        init_yaw = self.get_parameter('initial_pose_yaw').value

        # Particle Filter Initialization
        self.pf = ParticleFilter(
            min_particles=self.get_parameter('min_particles').value,
            max_particles=self.get_parameter('max_particles').value,
            initial_noise=[0.2, 0.2, 0.2]  # x, y, yaw noise
        )
        # 초기 파티클 생성
        self.pf.initialize(self.init_x, self.init_y, init_yaw)

        self.map_received = False
        self.last_odom_matrix = None

        # 추정 위치 캐싱 (업데이트가 없을 때도 TF를 보내기 위함)
        self.current_estimated_pose = [self.init_x, self.init_y, init_yaw]

        # TF Setup
        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)
        self.tf_broadcaster = TransformBroadcaster(self)

        # reliability를 RELIABLE로 해야지만 map정보를 받을 수 있습니다. 수정 금지
        map_qos = QoSProfile(
            depth=1,
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.TRANSIENT_LOCAL
        )

        # Map Subscription
        self.map_sub = self.create_subscription(
            OccupancyGrid,
            '/map',
            self.map_callback,
            map_qos
        )

        # Initial Pose Subscription (RViz "2D Pose Estimate")
        self.initial_pose_sub = self.create_subscription(
            PoseWithCovarianceStamped,
            '/initialpose',
            self.initial_pose_callback,
            10
        )

        # Laser Scan Subscription (Main Loop)
        self.scan_sub = self.create_subscription(
            LaserScan,
            '/scan',
            self.scan_callback,
            10
        )

        # Visualization Publishers
        self.pose_pub = self.create_publisher(PoseStamped, '/go1_pose', 10)
        self.particle_pub = self.create_publisher(PoseArray, '/particle_cloud', 10)

    def map_callback(self, msg):
        """지도를 수신하면 파티클 필터에 설정합니다."""
        self.get_logger().info(f'Received map: {msg.info.width}x{msg.info.height}, res: {msg.info.resolution}')
        self.pf.set_map(msg)
        self.map_received = True

    def initial_pose_callback(self, msg):
        """RViz 등에서 초기 위치를 재설정할 때 호출됩니다."""
        x = msg.pose.position.x
        y = msg.pose.position.y

        # Quaternion to Euler (Yaw)
        q = msg.pose.orientation
        _, _, yaw = tf_transformations.euler_from_quaternion([q.x, q.y, q.z, q.w])

        self.get_logger().info(f'Relocalizing to x:{x:.2f}, y:{y:.2f}, yaw:{yaw:.2f}')
        self.pf.initialize(x, y, yaw)
        self.last_odom_matrix = None  # 오도메트리 연속성 초기화

    def get_odom_pose(self, time):
        """특정 시간의 odom -> base_link TF를 구합니다."""
        try:
            trans = self.tf_buffer.lookup_transform('odom', 'base', time)
            return transform_to_matrix(trans.transform)
        except (LookupException, ConnectivityException, ExtrapolationException):
            try:
                # 실패 시 가장 최신 시간 조회
                trans = self.tf_buffer.lookup_transform('odom', 'base', Time())
                return transform_to_matrix(trans.transform)
            except Exception as e:
                self.get_logger().debug(f'TF lookup completely failed: {e}')
                return None

    def scan_callback(self, scan_msg):
        """
        메인 로직:
        1. Prediction (Odom Delta)
        2. Update (Likelihood Field)
        3. Resample
        4. TF Publish
        """
        if not self.map_received:
            return

        # 현재 스캔 시간
        current_time = Time.from_msg(scan_msg.header.stamp)

        # 1. Get Current Odom Pose
        curr_odom_matrix_3d = self.get_odom_pose(current_time)
        if curr_odom_matrix_3d is None:
            return

        # 네비게이션용 2D 오도메트리 생성 (Z, Roll, Pitch 제거)
        curr_odom_matrix_2d = get_2d_matrix(curr_odom_matrix_3d)

        # 업데이트 수행 여부 플래그
        do_update = False

        # 2. Prediction Step (Motion Model)
        if self.last_odom_matrix is not None:
            # 이전 프레임 기준 상대 변환 행렬 계산: T_prev_inv * T_curr
            # 이것은 Global(Odom) frame에서의 이동이므로, Robot frame에서의 dx, dy로 변환해야 함

            # 로봇 좌표계 기준 이동량 계산
            # T_delta = inv(T_prev) @ T_curr
            # T_delta는 로봇이 '이전 위치'에서 보았을 때 '현재 위치'가 어디인지 나타냄
            T_prev_inv = np.linalg.inv(self.last_odom_matrix)
            T_delta = T_prev_inv @ curr_odom_matrix_2d

            dx = T_delta[0, 3]
            dy = T_delta[1, 3]

            # 회전량 계산 (rotation matrix to euler)
            # T_delta[:3, :3] 은 회전 행렬
            _, _, dyaw = tf_transformations.euler_from_matrix(T_delta)

            # 이동량이 매우 작으면 필터 업데이트 스킵
            if abs(dx) > 0.001 or abs(dy) > 0.001 or abs(dyaw) > 0.001:
                # 파티클 필터 예측 단계 실행
                self.pf.predict(dx, dy, dyaw)
                do_update = True
                self.last_odom_matrix = curr_odom_matrix_2d
        else:
            # 첫 실행 시 초기화
            self.last_odom_matrix = curr_odom_matrix_2d


        if do_update:
            # 3. Update Step (Sensor Model)
            self.pf.update(
                scan_msg.ranges,
                scan_msg.angle_min,
                scan_msg.angle_increment,
                sensor_offset=[0.0, 0.0]  # base와 laser_link가 일치함. 확인 완료
            )
            self.pf.resample()
            self.current_estimated_pose = self.pf.get_estimated_pose()

        # 4. Get Estimated Pose (Map -> Base)
        estimated_pose = self.pf.get_estimated_pose()  # [x, y, yaw]

        # 5. Calculate Map -> Odom TF
        # T_map_to_odom = T_map_to_base * inv(T_odom_to_base)
        # 여기서 모든 계산은 2D 평면상에서 이루어져야 맵이 기울어지지 않음

        # T_map_to_base 생성
        q_est = tf_transformations.quaternion_from_euler(0, 0, estimated_pose[2])
        T_map_to_base = tf_transformations.quaternion_matrix(q_est)
        T_map_to_base[0, 3] = estimated_pose[0]
        T_map_to_base[1, 3] = estimated_pose[1]

        # T_odom_to_base의 역행렬 (반드시 2D화 된 매트릭스를 사용해야 함)
        T_base_to_odom = np.linalg.inv(curr_odom_matrix_2d)

        # T_map_to_odom 계산
        T_map_to_odom = T_map_to_base @ T_base_to_odom

        # Publish Map -> Odom TF
        self.publish_tf(T_map_to_odom, scan_msg.header.stamp)

        # 5. Publish
        self.publish_mcl_pose(estimated_pose, scan_msg.header.stamp)

        if do_update:
            self.publish_particles(scan_msg.header.stamp)

    def publish_tf(self, T, stamp):
        t = TransformStamped()
        t.header.stamp = stamp
        t.header.frame_id = 'map'
        t.child_frame_id = 'odom'

        translation = T[:3, 3]
        quat = tf_transformations.quaternion_from_matrix(T)

        t.transform.translation.x = float(translation[0])
        t.transform.translation.y = float(translation[1])
        t.transform.translation.z = 0.0 # Map->Odom은 2D 평면상 변환이어야 안전함
        t.transform.rotation.x = float(quat[0])
        t.transform.rotation.y = float(quat[1])
        t.transform.rotation.z = float(quat[2])
        t.transform.rotation.w = float(quat[3])

        self.tf_broadcaster.sendTransform(t)

    def publish_mcl_pose(self, pose_2d, stamp):
        msg = PoseStamped()
        msg.header.stamp = stamp
        msg.header.frame_id = 'map'
        msg.pose.position.x = float(pose_2d[0])
        msg.pose.position.y = float(pose_2d[1])
        msg.pose.position.z = self.init_z

        # 2D PF는 yaw만 추정하므로, roll/pitch는 0 혹은 IMU 값 사용해야 함.
        # 지침상 단순화를 위해 0 혹은 초기값 유지.
        q = tf_transformations.quaternion_from_euler(0, 0, float(pose_2d[2]))

        msg.pose.orientation.x = float(q[0])
        msg.pose.orientation.y = float(q[1])
        msg.pose.orientation.z = float(q[2])
        msg.pose.orientation.w = float(q[3])

        self.pose_pub.publish(msg)

    def publish_particles(self, stamp):
        # 성능을 위해 파티클을 다운샘플링하여 시각화할 수도 있음
        msg = PoseArray()
        msg.header.stamp = stamp
        msg.header.frame_id = 'map'

        particles = self.pf.particles
        # 시각화 부하를 줄이기 위해 최대 100개 정도만 퍼블리시 하거나 전체를 퍼블리시
        step = max(1, len(particles) // 10)

        for p in particles[::step]:
            pose = Pose()
            pose.position.x = float(p[0])
            pose.position.y = float(p[1])
            q = tf_transformations.quaternion_from_euler(0, 0, float(p[2]))
            pose.orientation.x = float(q[0])
            pose.orientation.y = float(q[1])
            pose.orientation.z = float(q[2])
            pose.orientation.w = float(q[3])
            msg.poses.append(pose)

        self.particle_pub.publish(msg)


if __name__ == '__main__':
    rclpy.init()
    node = GlobalLocalizerNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

File: src/localization/localization/global_localizer_node.cpp
#include <chrono>
#include <memory>
#include <string>
#include <vector>
#include <cmath>
#include <mutex>

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "geometry_msgs/msg/pose_array.hpp"
#include "geometry_msgs/msg/pose_with_covariance_stamped.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "nav_msgs/msg/occupancy_grid.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"

#include "tf2/LinearMath/Quaternion.h"
#include "tf2/LinearMath/Matrix3x3.h"
#include "tf2/LinearMath/Transform.h"
#include "tf2_ros/transform_broadcaster.h"
#include "tf2_ros/transform_listener.h"
#include "tf2_ros/buffer.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"

#include "particle_filter.hpp"

using namespace std::chrono_literals;

class GlobalLocalizerNode : public rclcpp::Node
{
public:
    GlobalLocalizerNode()
        : Node("global_localizer"), map_received_(false), first_odom_received_(false)
    {
        RCLCPP_INFO(this->get_logger(), "Global localizer (MCL) node initialized (C++)");

        // 1. Parameters
        this->declare_parameter("initial_pose_x", 0.0);
        this->declare_parameter("initial_pose_y", 1.0);
        this->declare_parameter("initial_pose_z", 0.5);
        this->declare_parameter("initial_pose_yaw", 0.0);
        this->declare_parameter("min_particles", 500);
        this->declare_parameter("max_particles", 2000);

        init_x_ = this->get_parameter("initial_pose_x").as_double();
        init_y_ = this->get_parameter("initial_pose_y").as_double();
        init_z_ = this->get_parameter("initial_pose_z").as_double();
        double init_yaw = this->get_parameter("initial_pose_yaw").as_double();
        int min_particles = this->get_parameter("min_particles").as_int();
        int max_particles = this->get_parameter("max_particles").as_int();

        // 2. Initialize Particle Filter
        // ParticleFilter 생성자의 파라미터에 맞춰 초기화
        pf_ = std::make_unique<ParticleFilter>(
            min_particles, max_particles, 
            0.2f, 0.2f, 0.2f); // init noise x, y, yaw

        pf_->initialize(static_cast<float>(init_x_), static_cast<float>(init_y_), static_cast<float>(init_yaw));

        // 3. TF Setup
        tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
        tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
        tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);

        // 4. QoS for Map (Reliable, Transient Local)
        rclcpp::QoS map_qos(rclcpp::KeepLast(1));
        map_qos.reliability(rclcpp::ReliabilityPolicy::Reliable);
        map_qos.durability(rclcpp::DurabilityPolicy::TransientLocal);

        // 5. Subscribers & Publishers
        map_sub_ = this->create_subscription<nav_msgs::msg::OccupancyGrid>(
            "/map", map_qos,
            std::bind(&GlobalLocalizerNode::map_callback, this, std::placeholders::_1));

        initial_pose_sub_ = this->create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
            "/initialpose", 10,
            std::bind(&GlobalLocalizerNode::initial_pose_callback, this, std::placeholders::_1));

        scan_sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
            "/scan", 10,
            std::bind(&GlobalLocalizerNode::scan_callback, this, std::placeholders::_1));

        pose_pub_ = this->create_publisher<geometry_msgs::msg::PoseStamped>("/go1_pose", 10);
        particle_pub_ = this->create_publisher<geometry_msgs::msg::PoseArray>("/particle_cloud", 10);
    }

private:
    // --- Member Variables ---
    std::unique_ptr<ParticleFilter> pf_;
    
    // State
    bool map_received_;
    bool first_odom_received_;
    
    double init_x_, init_y_, init_z_;
    
    // TF
    std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
    std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
    std::shared_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
    
    // Odom caching for motion model
    tf2::Transform last_odom_tf_;

    // ROS handles
    rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr map_sub_;
    rclcpp::Subscription<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr initial_pose_sub_;
    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr scan_sub_;
    rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr pose_pub_;
    rclcpp::Publisher<geometry_msgs::msg::PoseArray>::SharedPtr particle_pub_;

    // --- Helper Functions ---

    // 2D 평면(x, y, yaw)만 남기고 나머지(z, roll, pitch)를 제거한 Transform 반환
    tf2::Transform get_2d_transform(const tf2::Transform& src)
    {
        double x = src.getOrigin().x();
        double y = src.getOrigin().y();
        
        double roll, pitch, yaw;
        src.getBasis().getRPY(roll, pitch, yaw);

        tf2::Transform dest;
        tf2::Quaternion q;
        q.setRPY(0.0, 0.0, yaw);
        dest.setRotation(q);
        dest.setOrigin(tf2::Vector3(x, y, 0.0));
        return dest;
    }

    // --- Callbacks ---

    void map_callback(const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
    {
        RCLCPP_INFO(this->get_logger(), "Received map: %dx%d, res: %.3f", 
                    msg->info.width, msg->info.height, msg->info.resolution);
        
        // nav_msgs::OccupancyGrid의 data는 int8_t 배열이므로 바로 전달 가능
        // ParticleFilter::set_map 시그니처: 
        // void set_map(const std::vector<int8_t>& map_data, int width, int height, float resolution, float origin_x, float origin_y);
        
        pf_->set_map(msg->data, msg->info.width, msg->info.height, 
                     msg->info.resolution, 
                     msg->info.origin.position.x, 
                     msg->info.origin.position.y);
        
        map_received_ = true;
    }

    void initial_pose_callback(const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg)
    {
        double x = msg->pose.pose.position.x;
        double y = msg->pose.pose.position.y;
        
        tf2::Quaternion q;
        tf2::fromMsg(msg->pose.pose.orientation, q);
        double roll, pitch, yaw;
        tf2::Matrix3x3(q).getRPY(roll, pitch, yaw);

        RCLCPP_INFO(this->get_logger(), "Relocalizing to x:%.2f, y:%.2f, yaw:%.2f", x, y, yaw);
        
        pf_->initialize(static_cast<float>(x), static_cast<float>(y), static_cast<float>(yaw));
        
        // 오도메트리 연속성이 끊기므로 플래그 리셋
        first_odom_received_ = false;
    }

    void scan_callback(const sensor_msgs::msg::LaserScan::SharedPtr scan_msg)
    {
        if (!map_received_) return;

        rclcpp::Time current_time = scan_msg->header.stamp;

        // 1. Get Current Odom Pose (odom -> base)
        geometry_msgs::msg::TransformStamped odom_tf_msg;
        try {
            odom_tf_msg = tf_buffer_->lookupTransform(
                "odom", "base", 
                current_time, 
                rclcpp::Duration::from_seconds(0.1));
        } catch (tf2::TransformException &ex) {
            // 실패 시 최신 데이터 조회 시도
            try {
                odom_tf_msg = tf_buffer_->lookupTransform("odom", "base", rclcpp::Time(0));
            } catch (tf2::TransformException &ex2) {
                RCLCPP_DEBUG(this->get_logger(), "TF lookup failed: %s", ex2.what());
                return;
            }
        }

        tf2::Transform curr_odom_tf_3d;
        tf2::fromMsg(odom_tf_msg.transform, curr_odom_tf_3d);

        // 네비게이션용 2D 오도메트리 생성 (Z, Roll, Pitch 제거)
        tf2::Transform curr_odom_tf_2d = get_2d_transform(curr_odom_tf_3d);

        bool do_update = false;

        // 2. Prediction Step (Motion Model)
        if (first_odom_received_) {
            // T_delta = T_prev_inv * T_curr (Local movement in robot frame)
            tf2::Transform tf_delta = last_odom_tf_.inverse() * curr_odom_tf_2d;

            double dx = tf_delta.getOrigin().x();
            double dy = tf_delta.getOrigin().y();
            
            double droll, dpitch, dyaw;
            tf_delta.getBasis().getRPY(droll, dpitch, dyaw);

            // 이동량이 일정 이상일 때만 업데이트 수행
            if (std::abs(dx) > 0.0001 || std::abs(dy) > 0.0001 || std::abs(dyaw) > 0.0001) {
                pf_->predict(static_cast<float>(dx), static_cast<float>(dy), static_cast<float>(dyaw));
                do_update = true;
                last_odom_tf_ = curr_odom_tf_2d;
            }
        } else {
            // 첫 실행 시 초기화
            last_odom_tf_ = curr_odom_tf_2d;
            first_odom_received_ = true;
        }

        if (do_update) {
            // 3. Update Step (Sensor Model)
            // sensor_offset은 base와 laser 사이의 거리. 여기선 0으로 가정 (Python 코드와 동일)
            float sensor_offset[2] = {0.0f, 0.0f};
            
            pf_->update(scan_msg->ranges, 
                        scan_msg->angle_min, 
                        scan_msg->angle_increment, 
                        sensor_offset);
            
            pf_->resample();
        }

        // 4. Get Estimated Pose (Map -> Base)
        std::vector<float> est_pose = pf_->get_estimated_pose(); // [x, y, yaw]

        // 5. Calculate Map -> Odom TF
        // T_map_to_odom = T_map_to_base * (T_odom_to_base)^-1
        
        // 5-1. T_map_to_base 생성
        tf2::Transform tf_map_to_base;
        tf2::Quaternion q_est;
        q_est.setRPY(0, 0, est_pose[2]);
        tf_map_to_base.setRotation(q_est);
        tf_map_to_base.setOrigin(tf2::Vector3(est_pose[0], est_pose[1], 0.0));

        // 5-2. T_odom_to_base의 역행렬 = T_base_to_odom
        // 주의: 반드시 2D화 된 오도메트리를 사용해야 맵이 기울어지지 않음
        tf2::Transform tf_base_to_odom = curr_odom_tf_2d.inverse();

        // 5-3. 최종 T_map_to_odom
        tf2::Transform tf_map_to_odom = tf_map_to_base * tf_base_to_odom;

        // Publish TF
        publish_tf(tf_map_to_odom, scan_msg->header.stamp);

        // 6. Publish Pose & Particles
        publish_mcl_pose(est_pose, scan_msg->header.stamp);

        if (do_update) {
            publish_particles(scan_msg->header.stamp);
        }
    }

    void publish_tf(const tf2::Transform& transform, const rclcpp::Time& stamp)
    {
        geometry_msgs::msg::TransformStamped t;
        t.header.stamp = stamp;
        t.header.frame_id = "map";
        t.child_frame_id = "odom";

        t.transform = tf2::toMsg(transform);
        
        // Map->Odom은 2D 평면상 변환이어야 하므로 Z는 강제로 0 (이미 로직상 0이지만 확실히)
        t.transform.translation.z = 0.0;

        tf_broadcaster_->sendTransform(t);
    }

    void publish_mcl_pose(const std::vector<float>& pose_2d, const rclcpp::Time& stamp)
    {
        geometry_msgs::msg::PoseStamped msg;
        msg.header.stamp = stamp;
        msg.header.frame_id = "map";
        
        msg.pose.position.x = pose_2d[0];
        msg.pose.position.y = pose_2d[1];
        msg.pose.position.z = init_z_; // 초기 설정된 Z값 유지

        tf2::Quaternion q;
        q.setRPY(0, 0, pose_2d[2]);
        msg.pose.orientation = tf2::toMsg(q);

        pose_pub_->publish(msg);
    }

    void publish_particles(const rclcpp::Time& stamp)
    {
        // 구독자가 없으면 연산조차 하지 않음
        if (particle_pub_->get_subscription_count() == 0) return;

        // 성능을 위해 파티클 다운샘플링 (Python 코드와 유사하게 1/10 정도)
        const auto& particles = pf_->get_particles();
        size_t step = std::max(size_t(1), particles.size() / 10);

        geometry_msgs::msg::PoseArray msg;
        msg.header.stamp = stamp;
        msg.header.frame_id = "map";
        
        msg.poses.reserve(particles.size() / step + 1);

        for (size_t i = 0; i < particles.size(); i += step) {
            geometry_msgs::msg::Pose pose;
            pose.position.x = particles[i].x;
            pose.position.y = particles[i].y;
            pose.position.z = 0.0; // 파티클은 2D

            tf2::Quaternion q;
            q.setRPY(0, 0, particles[i].yaw);
            pose.orientation = tf2::toMsg(q);

            msg.poses.push_back(pose);
        }

        particle_pub_->publish(msg);
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<GlobalLocalizerNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
